### PR TITLE
Fix #297 : fix corruption in startxref or xref table

### LINKED ISSUE #297: ValueError: invalid literal for int() with base 10: 'n'
I'm trying to read a pdf file (of the simple pdf file)
but can't read pdf file: (Acrobat Reader program is open as normal.)

run pypdf2-test.py

```console
$ python --version
Python 2.6.6

$ python ./pypdf2-test.py
Traceback (most recent call last):
  File "./zz.py", line 3, in <module>
    infile = PdfFileReader('source.pdf', 'rb')
  File "/usr/lib/python2.6/site-packages/PyPDF2/pdf.py", line 1148, in __init__
    self.read(stream)
  File "/usr/lib/python2.6/site-packages/PyPDF2/pdf.py", line 1867, in read
    idnum, generation = self.readObjectHeader(stream)
  File "/usr/lib/python2.6/site-packages/PyPDF2/pdf.py", line 1731, in readObjectHeader
    return int(idnum), int(generation)
ValueError: invalid literal for int() with base 10: 'n'
```

$ cat ./pypdf2-test.py

```python
from` PyPDF2 import PdfFileWriter, PdfFileReader

infile = PdfFileReader('source.pdf', 'rb')
output = PdfFileWriter()

for i in xrange(infile.getNumPages()):
    p = infile.getPage(i)
    if p.getContents(): # getContents is None if  page is blank
        output.addPage(p)

with open('newfile.pdf', 'wb') as f:
   output.write(f)
```

$ cat ./source.pdf (sample pdf file)

```text
%PDF-1.1

1 0 obj
<<
 /Type /Catalog
 /Outlines 2 0 R
 /Pages 3 0 R
 /OpenAction 7 0 R
>>
endobj

2 0 obj
<<
 /Type /Outlines
 /Count 0
>>
endobj

3 0 obj
<<
 /Type /Pages
 /Kids [4 0 R]
 /Count 1
>>
endobj

4 0 obj
<<
 /Type /Page
 /Parent 3 0 R
 /MediaBox [0 0 612 792]
 /Contents 5 0 R
 /Resources <<
             /ProcSet [/PDF /Text]
             /Font << /F1 6 0 R >>
            >>
>>
endobj

5 0 obj
<< /Length 56 >>
stream
BT /F1 12 Tf 100 700 Td 15 TL (test example) Tj ET
endstream
endobj

6 0 obj
<<
 /Type /Font
 /Subtype /Type1
 /Name /F1
 /BaseFont /Helvetica
 /Encoding /MacRomanEncoding
>>
endobj

7 0 obj
<<
 /Type /Action
 /S /JavaScript
 /JS (app.alert({cMsg: 'Hello alert', cTitle: 'Testing PDF', nIcon: 3});)
>>
endobj

xref
0 8
0000000000 65535 f
0000000012 00000 n
0000000109 00000 n
0000000165 00000 n
0000000234 00000 n
0000000439 00000 n
0000000553 00000 n
0000000677 00000 n
trailer
<<
 /Size 8
 /Root 1 0 R
>>
startxref
842
%%EOF
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
Proposal for #297

    Issue is due to corrupted startxref pointer and/or xref table
    Fixed proposed if incorrect pointer look for all objs to rebuild an xref table

    Tests updated

### CODE REVIEW

**Comment by MartinThoma** [comment]
#297

File: Tests/test_reader.py [hunk 1]
```diff
@@ -214,16 +214,11 @@ def test_get_images_raw(strict, with_prev_0, should_fail):
...
-        "Help would be appreciated!"
-    )
-)
 def test_issue297():
     path = os.path.join(RESOURCE_ROOT, "issue-297.pdf")
-    reader = PdfFileReader(path, "rb")
-    reader.getPage(0)
+    with pytest.raises(UserWarning):
+        reader = PdfFileReader(path, "rb")
+        reader.getPage(0)
```
Comment (MartinThoma) [inline Tests/test_reader.py:221]: This deactivates the test. At least for `strict=False` I would expect `reader.extractText()` to give something reasonable.

The rest of the PR might be fine, but it does not fix issue297. I will not merge this part.

**Comment by pubpub-zz** [comment]
@MartinThoma 
Thanks for your review. I do not master pytest. Test reviewed to check both Strict True and False.

**Comment by MartinThoma** [comment]
I understand :-)

This might help to capture both cases:

```
@pytest.mark.parametrize(
    "set_strict, [True, False]
)
def test_issue297(set_strict):
    .... # the test: you pass set_strict to the PdfFileReader and you only do 'with pytest.raises' if set_strict is True
```

**Comment by pubpub-zz** [comment]
@MartinThoma ,
 The expected result not being the same I've included both in the test manually isn't this good for you ?
```
def test_issue297():
    path = os.path.join(RESOURCE_ROOT, "issue-297.pdf")
    with pytest.raises(PdfReadWarning) as exc:
        print(exc)
        reader = PdfFileReader(path,strict=True)
        reader.getPage(0)
    assert ( exc.value.args[0].find("startxref") > 0)
    reader = PdfFileReader(path,strict=False)
    reader.getPage(0)

```

**Comment by MartinThoma** [comment]
@pubpub-zz Thank your for pointing it out - yes, that is ok :-) The parametrize option is better as it shows clearer where an issue is (if there was one) + it avoids a bit of code duplication. But leave it as is - that is just a nitpick; I'll polish that in another commit. Thank you :hugs: 

I'll not merge it for now. In my roof apartment one of the windows just broke out ... I'm still shocked + cannot concentrate. I'll likely do review it again tomorrow when I can focus again :sweat_smile:

**Comment by pubpub-zz** [comment]
@MartinThoma, good luck with your Window(...s 10 or 11 ? That is the most common window which is breaking down 🤣)

**Comment by MartinThoma** [comment]
Good morning!

There seems to be something off with `test_read_unknown_zero_pages`. You've commented, that this is necessary due to the double-percentage in the beginning (`%%PDF-1.7`). However, this is a part of the percentage formatting. The actual formatted byte-string looks like this:

```pdf
%PDF-1.7
1 0 obj << /Count 1 /Kids [4 0 R] /Type /Pages >> endobj
2 0 obj << >> endobj
3 0 obj << >> endobj
4 0 obj << /Contents 3 0 R /CropBox [0.0 0.0 2550.0 3508.0] /MediaBox [0.0 0.0 2550.0 3508.0] /Parent 1 0 R /Resources << /Font << >> >> /Rotate 0 /Type /Page >> endobj
5 0 obj << /Pages 0 0 R /Type /Catalog >> endobj
xref 1 5
0000000010 00000 n
0000000067 00000 n
0000000088 00000 n
0000000109 00000 n
0000000278 00000 n
trailer << /Root 5 1 R /Size 6 >>
startxref 327
%%EOF'
```

As you can see, the byte string starts with `%PDF-1.7` and not `%%PDF-1.7`.

Can you help me understand why the `-1` is now necessary?

**Comment by MartinThoma** [comment]
(minor comment, feel free to ignore: Running [black](https://pypi.org/project/black/) over just the test is fine: `black Tests/test_reader.py`. Please don't run it over the complete codebase as this would add to many changes, but I already had it running over the Tests. For this reason it will just beautify your additions)

File: PyPDF2/pdf.py [hunk 2]
```diff
@@ -1854,6 +1854,35 @@ def read(self, stream):
             if line[:9] != b_("startxref"):
                 raise PdfReadError("startxref not found")
 
+        #check and eventually correct the startxref only in not strict
+        rebuildXrefTable = False
+        try:
+            stream.seek(startxref - 1,0) #-1 to check character before
+            line=stream.read(1)
+            if line not in b_("\r\n \t"):
```
Comment (MartinThoma) [inline PyPDF2/pdf.py:1862]: I would expect `x` as in `xref`. Why do you check fir a whitespace?

Comment (MartinThoma) [reply PyPDF2/pdf.py:1862]: uuuh, wait, do you want to ensure that it's not by accient xref, e.g. within a text string? Might this cause other issues? Is it actually guaranteed that there has to be a whitespace?

**Comment by pubpub-zz** [comment]
> As you can see, the byte string starts with `%PDF-1.7` and not `%%PDF-1.7`.
> 
> Can you help me understand why the `-1` is now necessary?

due to the formatting, the original formatting string includes a double '%%' that is converted to a single during formating. the point is that the .find(b"xref") applies the search on the original function. therefore the 327 points to the r and not the x of the xref.
the original code was not checking for the xref not reporting any error being tolerant to 'invalid' pdf data.
with the new code, strict=true will reject the file - not compliant with the spec - but with string=false, the xref table will be rebuilt, being able to accept the file
@MartinThoma , Do you think it would be better to accept this pointer error as it used to be without rebuilding the xref table?

File: PyPDF2/pdf.py (hunk 2 above)
Comment (pubpub-zz) [reply PyPDF2/pdf.py:1862]: wanted to ensure that the xref is a word not something else

**Comment by MartinThoma** [comment]
> the .find(b"xref") applies the search on the original

Thank you - that was what I missed :pray: :+1:  Thank you for fixing the tests!

> Do you think it would be better to accept this pointer error as it used to be without rebuilding the xref table?

I fear that we might now reject / fail to parse a lot of PDFs that we could parse beforehand. Fixing the xref table if we can is pretty nice - and raising an exception if it's broken as well. The part that I'm worried about is broken PDFs, that we could still parse before.

**Comment by pubpub-zz** [comment]
> I fear that we might now reject / fail to parse a lot of PDFs that we could parse beforehand. Fixing the xref table if we can is pretty nice - and raising an exception if it's broken as well 👍
> The part that I'm worried about is broken PDFs, that we could still parse before.

So did I. I've extended to test to cover both nominal case and incorrect pointer. A loop has been detected and fixed in such cases.

File: PyPDF2/pdf.py [hunk 3]
```diff
@@ -1843,6 +1843,35 @@ def read(self, stream):
...
+                raise PdfReadWarning("incorrect startxref pointer(1)",line)
```
Comment (MartinThoma) [inline PyPDF2/pdf.py:1852]: Warnings should not be raised, but issued with the warnings library:

```suggestion
                warnings.warn("incorrect startxref pointer(1): {}".format(line), PdfReadWarning)
```

File: Tests/test_reader.py [hunk 4]
```diff
@@ -200,29 +204,36 @@ def test_get_images_raw(strict, with_prev_0, should_fail):
         pdf_data.find(b"4 0 obj"),
         pdf_data.find(b"5 0 obj"),
         b"/Prev 0 " if with_prev_0 else b"",
-        pdf_data.find(b"xref"),
+        # startx_correction should be -1 due to double % at the beginning indiducing an error on startxref computation
+        pdf_data.find(b"xref") + startx_correction,
     )
+    print((strict, with_prev_0, should_fail))
+    print(pdf_data)
```
Comment (MartinThoma) [inline Tests/test_reader.py:211]: Print statements should not remain in the library.

File: Tests/test_reader.py [hunk 5]
```diff
@@ -200,29 +204,36 @@ def test_get_images_raw(strict, with_prev_0, should_fail):
...
     pdf_stream = io.BytesIO(pdf_data)
     if should_fail:
-        with pytest.raises(PdfReadError) as exc:
+        with pytest.raises(Exception) as exc:
             PdfFileReader(pdf_stream, strict=strict)
-        assert (
-            exc.value.args[0]
-            == "/Prev=0 in the trailer (try opening with strict=False)"
-        )
+        print(exc.type)
```
Comment (MartinThoma) [inline Tests/test_reader.py:216]: Another print call

File: Tests/test_reader.py [hunk 6]
```diff
@@ -200,29 +204,36 @@ def test_get_images_raw(strict, with_prev_0, should_fail):
...
-    reason=(
-        "It's still broken - and unclear what the issue is. "
-        "Help would be appreciated!"
-    )
-)
 def test_issue297():
     path = os.path.join(RESOURCE_ROOT, "issue-297.pdf")
-    reader = PdfFileReader(path, "rb")
+    with pytest.raises(PdfReadWarning) as exc:
+        print(exc)
```
Comment (MartinThoma) [inline Tests/test_reader.py:232]: and another one

File: Tests/test_reader.py [hunk 7]
```diff
@@ -200,29 +204,36 @@ def test_get_images_raw(strict, with_prev_0, should_fail):
...
+        reader = PdfFileReader(path, strict=True)
+        reader.getPage(0)
+    assert exc.value.args[0].find("startxref") > 0
```
Comment (MartinThoma) [inline Tests/test_reader.py:235]:
```suggestion
    assert "startxref" in exc.value.args[0]
```

**Comment by MartinThoma** [comment]
Thank you very much for all the time and effort you've put into this! I appreciate it!

**Comment by MartinThoma** [comment]
@pubpub-zz I've created a follow-up PR that improves this IMHO a little bit. Nothing substantial, but from a maintainers perspective important. I'll add the reasoning in a second: https://github.com/py-pdf/PyPDF2/pull/830

**Pull request merged**